f/common_logic/tests/data_validation_test.py

```diff
@@ -0,0 +1,86 @@
+from pathlib import Path
+
+import pandas as pd
+
+from f.common_logic.data_validation import detect_structured_data_type
+
+
+def test_structured_data_type__valid_gpx(tmp_path: Path):
+    file_path = tmp_path / "test.gpx"
+    file_path.write_text(
+        "<?xml version='1.0' encoding='UTF-8'?><gpx version='1.1' creator='gpxpy' xmlns='http://www.topografix.com/GPX/1/1'><trk><trkseg><trkpt lat='48.8584' lon='2.3584'><ele>100</ele></trkpt></trkseg></trk></gpx>"
+    )
+    assert detect_structured_data_type(file_path) == "gpx"
+
+
+def test_structured_data_type__valid_kml(tmp_path: Path):
+    file_path = tmp_path / "test.kml"
+    file_path.write_text(
+        "<?xml version='1.0' encoding='UTF-8'?><kml xmlns='http://www.opengis.net/kml/2.2'><Document><Folder><Placemark><Point><coordinates>2.3584,48.8584</coordinates></Point></Placemark></Folder></Document></kml>"
+    )
+    assert detect_structured_data_type(file_path) == "kml"
+
+
+def test_structured_data_type__valid_csv(tmp_path: Path):
+    file_path = tmp_path / "test.csv"
+    file_path.write_text("col1,col2\nval1,val2")
+    assert detect_structured_data_type(file_path) == "csv"
+
+
+def test_structured_data_type__invalid_csv(tmp_path: Path):
+    file_path = tmp_path / "test.csv"
+    file_path.write_text("key:value,value,more\n1:2:3\nnot,really,a,csv")
+    assert not detect_structured_data_type(file_path) == "csv"
+
+
+def test_structured_data_type__valid_json(tmp_path: Path):
+    file_path = tmp_path / "test.json"
+    file_path.write_text(
+        """
+        {
+            "key": "value"
+        }
+        """
+    )
+    assert detect_structured_data_type(file_path) == "json"
+
+
+def test_structured_data_type__malformed_json(tmp_path: Path):
+    file_path = tmp_path / "bad.json"
+    file_path.write_text("{ this is not valid json }")
+    assert detect_structured_data_type(file_path) == "unsupported"
```

Comment:
nitpick: I feel this function ought to raise an exception (let's call it `UnrecognizedDataType`). Else it looks like "unsupported" is the name of a valid data type.

It's ok if you reject this idea.

Reply:
Similar to my comment about [error handling](https://github.com/ConservationMetrics/gc-scripts-hub/pull/110#discussion_r2174034313) in #110, I need to take the time to recall how this can work across a Windmill app.

If this wasn't written with usage in a Windmill app in mind, I would totally agree with this. And it does raise a question around how somebody might use this "common logic" function in a different context.

I'm going to make an explicit note to revisit this as part of the work proposed in https://github.com/ConservationMetrics/gc-scripts-hub/issues/73.